### DISCUSSION TITLE
Ability to use custom database functions in filters

### DISCUSSION THREAD

**Discussion by kostopravby** [directus/directus#8694] [Archive]
I want to use Directus for storing multipolygons and check if some point is inside polygon.
I didn't find any filter from https://docs.directus.io/reference/filter-rules/. Then I created custom db function which looks like this
```
CREATE OR REPLACE FUNCTION public.checkgeo(geometry, varchar)
	RETURNS boolean 	
AS $$
	begin
		RETURN EXISTS (
		select 1 from test t where ST_contains($1,ST_GeomFromText(concat('POINT(',$2,')'),4326)));
	END
     
$$ LANGUAGE plpgsql;

```
and sql query 
`select t.discount from test t where checkgeo(t.geo,'23.097920215411235 11.84472406572692');`
works well.
Now I want to use that function in REST API (items/test?filter[geo][_chekgeo]=xxxx) but this don't work. It always returns all items from collection.
Is there any way to use custom filters or existing filters with geo data?

**Comment by joselcvarela** [comment]
Hello @kostopravby,
Extending filters is not possible and I don't know if that makes sense. Probably we could create a 'is inside polygon' filter into the core

But for now you could try something like `items/test?filter[geo][checkgeo(t.geo,'23.097920215411235 11.84472406572692')]=1` (not sure if it will work)

**Reply by kostopravby** [reply]
unfortunately this does not work with any arguments in checkgeo even with empty value. I will check db query log
will be glad for any other solution. maybe this helps https://github.com/substack/point-in-polygon

**Reply by azrikahar** [reply]
@kostopravby Never personally try this so I'm not sure if it'll work, but perhaps you can try using a custom endpoint and use `knex.raw` to pass the arguments into your postgresql function as mentioned here: https://github.com/knex/knex/issues/185

Note: Knex instance is available for custom endpoints under `context.database`: https://docs.directus.io/guides/api-endpoints/#_2-develop-your-custom-endpoint. Another useful reference for it's usage from another user: https://github.com/directus/directus/discussions/3312#discussioncomment-147883

**Comment by rijkvanzanten** [comment]
> Is there any way to use custom filters or existing filters with geo data?

No, there's no way to inject/run SQL into/from any of the API endpoints.

We can look into exposing `ST_Within` as an API filter for geometry type. @Oreilles do you happen to know if that's generally available in all database vendors?

**Comment by oreilles** [comment]
It is apparently not documented yet in the link you provided, but the API does support `intersects` and `nintersects` geometric filter operators, which both accepts any GeoJSON as an argument.

So you should be able to fetch `items/test?filter[geo][_intersects]={"type":"Point","coordinates":[23.1, 11.8]}`

**Reply by rijkvanzanten** [reply]
Added those missing docs in https://github.com/directus/directus/pull/8742

**Reply by kostopravby** [reply]
@Oreille With 9.0.1 all works fine. Just upgraded to 9.4.2 and get 
`{"errors":[{"message":"\"_intersects\" has to be a valid GeoJSON object","extensions":{"code":"INVALID_QUERY"}}]}`
example: `coverage?filter[geo][_intersects]={"type": "Point","coordinates":[3.0981523,41.8428085]}`

**Reply by oreilles** [reply]
@kostopravby  Please submit a bug report so that I can track that one !

**Reply by kostopravby** [reply]
done [10869](https://github.com/directus/directus/issues/10869)
